### PR TITLE
Exclude export netsuite UI

### PR DESCRIPTION
**Pull request by anishfyle**

### CODE REVIEW

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+      };
+      this.settingsService
+        .postSkipExport(that.workspaceId, payload1)
+        .subscribe((skipExport: SkipExport) => {});
+    }
+  }
+
+  setOperatorField(conditionField: string) {
+    const operatorList = [];
+    if (conditionField === 'claim_number') {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:165]: conditionField === 'claim_number' || conditionField === 'employee_email' || conditionField === 'report_title'

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+          conditionSelected.is_custom
+        );
+      }
+    );
+  }
+
+  conditionFieldWatcher_1() {
+    this.skipExportForm.controls.condition_1.valueChanges.subscribe(
+      (conditionSelected) => {
+        this.skipExportForm.controls.value_1.reset();
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:237]: most of the logic is same as `conditionFieldWatcher1`, please make a common function and reuse

File: src/app/core/services/mappings.service.ts [hunk 3]
```diff
@@ -92,15 +93,20 @@ export class MappingsService {
     return this.apiService.get(`/workspaces/${workspaceId}/netsuite/netsuite_fields/`, {});
   }
 
+  getSkipExportConditionField(): Observable<ExpenseField[]> {
```
Comment (ashwin1111) [inline src/app/core/services/mappings.service.ts:96]: getFyleCustomFields

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+
+  setConditionField(conditionValue) {
+    this.conditionField = conditionValue;
+  }
+
+  getCustomConditions() {
+    this.mappingsService
+      .getSkipExportConditionField()
+      .subscribe((conditionValue) => {
+        this.setConditionField(conditionValue);
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:348]: directly assign value to `this.conditionField`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    return false;
+  }
+
+  getAllSettings() {
+    this.fieldWatcher();
+    this.getCustomConditions();
+  }
+
+  selectedOptions() {
+    this.settingsService.getSkipExport(2).subscribe((skipExport) => {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:365]: remove 2 and add workspace id

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+      this.data = skipExport.results[0];
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:366]: assign to const

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+      this.data1 = skipExport.results[1];
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:367]: same

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+      let ofType = '';
+      if (this.data.condition === 'employee_email') {
+        ofType = 'SELECT';
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:370]: don't we get this info from the API?

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+  payload = {};
+  joinBy = [{ value: 'AND' }, { value: 'OR' }];
+
+  ngOnInit() {
+    this.condition = false;
+    this.addConditionButton = true;
+    this.initFormGroups();
+    this.getAllSettings();
+    this.selectedOptions();
+  }
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:52]: we'll need to implement isLoading

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: src/app/core/models/general-setting.model.ts [hunk 10]
```diff
@@ -14,6 +14,7 @@ export type GeneralSetting = {
     auto_map_employees: string;
     employee_field_mapping: string;
     memo_structure?: string[];
+    // skip_export?: string[];
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/general-setting.model.ts]: have removed the commented code.

File: src/app/core/models/skip-export-selected.model.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+/* tslint:disable */
+
+export type SkipExportSelected = {
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export-selected.model.ts:3]: - changed to ExpenseFilterResponse.

- file name changed from skip-export-selected.model.ts ->  **TO**  -> expense-filter-response.model.ts

- have changed imports and name in setting service.

File: src/app/core/models/general-setting.model.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/general-setting.model.ts]: - sync-up

File: src/app/core/models/skip-export-selected.model.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  count: number;
+  results: [
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export-selected.model.ts:5]: - done ✅

File: src/app/core/models/skip-export.model.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/* tslint:disable */
+
+export type SkipExport = {
+  condition: string;
+  operator: string;
+  values: Array<any>;
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export.model.ts:6]: - done ✅

File: src/app/core/models/skip-export.model.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+  rank: number;
+  join_by: string;
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export.model.ts:8]: - sync-up

File: src/app/core/models/skip-export.model.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+/* tslint:disable */
+
+export type SkipExport = {
+  condition: string;
+  operator: string;
+  values: Array<any>;
+  rank: number;
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export.model.ts:7]: - sync-up

File: src/app/core/models/skip-export.model.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+/* tslint:disable */
+
+export type SkipExport = {
+  condition: string;
+  operator: string;
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/models/skip-export.model.ts:5]: - sync-up

File: src/app/core/services/settings.service.ts [hunk 17]
```diff
@@ -120,6 +124,17 @@ export class SettingsService {
...
   }
 
+  @CacheBuster({
+    cacheBusterNotifier: skipExportCache
+  })
+  postSkipExport(workspaceId: number, skipExport: SkipExport): Observable<SkipExport> {
+    workspaceId = this.workspaceService.getWorkspaceId();
+    return this.apiService.post(`/workspaces/${workspaceId}/fyle/expense_filters/`, skipExport);
+  }
+  getSkipExport(workspaceId: number): Observable<SkipExportSelected> {
```
[the remark this answers is not in the document]
Comment (anishfyle) [reply src/app/core/services/settings.service.ts]: - done ✅

File: src/app/core/services/mappings.service.ts (hunk 3 above)
Comment (anishfyle) [reply src/app/core/services/mappings.service.ts:96]: - done

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+  joinByOptions = [{ value: 'AND' }, { value: 'OR' }];
+
+  resetOptions() {
+    this.skipExportForm.controls.join_by.reset();
+    this.skipExportForm.controls.condition2.reset();
+    this.skipExportForm.controls.operator2.reset();
+    this.skipExportForm.controls.value2.reset();
+  }
+
+  addCondition() {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:53]: `updateAdditionalFilterVisibility()`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+          };
+          const type = this.conditionFieldOptions.filter(
+            (fieldOption) => fieldOption.field_name === element.condition
+          )[0].type;
+          selectedConditionOption.type = type;
+          conditionArray.push(selectedConditionOption);
+        });
+
+        if (conditionArray.length > 1 && formOptions1.join_by != null) {
+          this.addCondition();
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:362]: `updateAdditionalFilterVisibility()`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+    return {
+      claim_number: operatorList,
+      employee_email: operatorList,
+      spent_at: operatorList,
+      report_title: operatorList,
+    }[conditionField];
+  }
+
+  setCustomOperatorOptions(isCustomCondition, rank) {
+    if (isCustomCondition && rank === 1) {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:224]: this function is called only when is_custom is true, so remove the arg off, not needed

Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:224]: rank: number

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 21]
```diff
@@ -0,0 +1,459 @@
...
+      this.operatorFieldOptions1 = [
+        {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:226]: assign the options to a global variable, and copy the value in this block

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 22]
```diff
@@ -0,0 +1,459 @@
...
+          value: 'is_empty',
+        },
+        {
+          label: 'is not empty',
+          value: 'is_not_empty',
+        },
+      ];
+    } else if (isCustomCondition && rank === 2) {
+      this.operatorFieldOptions2 = [
+        {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:241]: this.operatorFieldOptions2 = this.operatorFieldDefaultOptions.filter();

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 23]
```diff
@@ -0,0 +1,459 @@
...
+        value: 'lte',
+        label: 'is it on or before',
+      });
+    } else if (conditionField === 'report_title') {
+      operatorList.push({
+        value: 'icontains',
+        label: 'contains',
+      });
+    }
+    return {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:215]: `return operatorList;`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 24]
```diff
@@ -0,0 +1,459 @@
...
+          selectedConditionOption.type = type;
+          conditionArray.push(selectedConditionOption);
+        });
+
+        if (conditionArray.length > 1 && formOptions1.join_by != null) {
+          this.addCondition();
+        }
+
+        formOptions1.is_custom
+          ? this.setCustomOperatorOptions(
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:366]: do this in if else block

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 25]
```diff
@@ -0,0 +1,459 @@
...
+            is_custom: element.is_custom,
+          };
+          const type = this.conditionFieldOptions.filter(
+            (fieldOption) => fieldOption.field_name === element.condition
+          )[0].type;
+          selectedConditionOption.type = type;
+          conditionArray.push(selectedConditionOption);
+        });
+
+        if (conditionArray.length > 1 && formOptions1.join_by != null) {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:361]:
```
if (conditionArray.length) {
 // line 365-372

 if (conditionArray.length > 1) {
    // same option setting logics

    updateAdditionalFilterVisibility()
 }
}
```

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 26]
```diff
@@ -0,0 +1,459 @@
...
+                formOptions2.is_custom,
+                formOptions2.rank
+              )
+            : (this.operatorFieldOptions2 = this.setDefaultOperatorOptions(
+                formOptions2.condition
+              ));
+        }
+
+        const operatorSelected1 = formOptions1.operator === 'isnull' && formOptions1.values[0] === 'True' ?  'is_empty' : formOptions1.operator === 'isnull' && formOptions1.values[0] === 'False' ? 'is_not_empty' : formOptions1.operator;
+        const operatorSelected2 = formOptions2.operator === 'isnull' && formOptions2.values[0] === 'True' ?  'is_empty' : formOptions2.operator === 'isnull' && formOptions2.values[0] === 'False' ? 'is_not_empty' : formOptions2.operator;
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:385]: `selectedOperator1`

Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:385]:
```
        const condition1 = conditionArray.length > 0 ? conditionArray[0] : '';
        let operator1;
        if (condition1.operator === 'isnull') {
          if (condition1.values[0] === 'True') {
            operator1 = 'is_empty';
          } else {
            operator1 = 'is_not_empty';
          }
        } else {
          operator1 = condition1.operator;
        }
```

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 27]
```diff
@@ -0,0 +1,459 @@
...
+        this.skipExportForm = this.formBuilder.group({
+          condition1: [
+            conditionArray.length > 0 ? conditionArray[0] : '',
+            [Validators.required],
+          ],
+          operator1: [operatorSelected1, [Validators.required]],
+          value1: [
+            conditionArray[0].type === 'DATE'
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:393]: refactor similar to operator

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 28]
```diff
@@ -0,0 +1,459 @@
...
+            formOptions1.join_by != null
+              ? [formOptions2.custom_field_type]
+              : [''],
+          searchOption: [''],
+        });
+        this.fieldWatcher();
+        this.isLoading = false;
+      },
+      () => {
+        this.getCustomConditions();
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:433]: this block is not needed, API won't fail

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 29]
```diff
@@ -0,0 +1,459 @@
...
+
+  getAllSettings() {
+    forkJoin([
+      this.mappingsService.getFyleCustomFields(),
+      this.settingsService.getSkipExport(this.workspaceId),
+    ]).subscribe(
+      (responses) => {
+        this.conditionFieldOptions = responses[0];
+        const formOptions1 = responses[1].results[0];
+        const formOptions2 = responses[1].results[1];
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:346]: remove

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 30]
```diff
@@ -0,0 +1,459 @@
...
+        this.skipExportForm.controls.value2.reset();
+        if (this.showSecondFilter === true) {
+          if (conditionSelected.is_custom === true) {
+            this.setCustomOperatorOptions(conditionSelected.is_custom, 2);
+          } else if (conditionSelected.is_custom === false) {
+            this.operatorFieldOptions2 = this.setDefaultOperatorOptions(
+              conditionSelected.field_name
+            );
+          }
+        }
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:283]: `this.resetFields(condition1, value1, conditionSelected);`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 31]
```diff
@@ -0,0 +1,459 @@
...
+        {
+          label: 'is not empty',
+          value: 'is_not_empty',
+        },
+      ];
+    }
+  }
+
+  conditionFieldWatcher() {
+    this.skipExportForm.controls.condition1.valueChanges.subscribe(
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:258]:
```
resetFields(operator, value, conditionSelected) {
    this.skipExportForm.controls[operator].reset();
    this.skipExportForm.controls[value].reset();
    if (conditionSelected.is_custom === true) {
      this.setCustomOperatorOptions(conditionSelected.is_custom, 1);
    } else if (conditionSelected.is_custom === false) {
      this.operatorFieldOptions1 = this.setDefaultOperatorOptions(
        conditionSelected.field_name
      );
    }
  }

  ```

Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:258]: if rank is 1 => set value to operatorFieldOptions1, else set value to operatorFieldOptions2

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 32]
```diff
@@ -0,0 +1,459 @@
...
+    this.skipExportForm.controls.operator2.reset();
+    this.skipExportForm.controls.value2.reset();
+  }
+
+  addCondition() {
+    this.showSecondFilter = true;
+    this.showConditionButton = false;
+  }
+
+  remCondition() {
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:58]: send true to false as arg to updateAdditionalFilterVisibility() and set value to boolean variables

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts (hunk 18 above)
Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:53]: `showAdditionalCondition: bool`

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts (hunk 25 above)
Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:361]: if (conditionArray.length) {
 // line 365-372

 if (conditionArray.length > 1) {
    // same option setting logics

    updateAdditionalFilterVisibility()
 }
}

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 33]
```diff
@@ -0,0 +1,459 @@
...
+          condition1: [
+            conditionArray.length > 0 ? conditionArray[0] : '',
+            [Validators.required],
+          ],
+          operator1: [operatorSelected1, [Validators.required]],
+          value1: [
+            conditionArray[0].type === 'DATE'
+              ? new Date(formOptions1.values[0])
+              : formOptions1.values,
+            [Validators.required],
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:396]: do this conditionally

Comment (ashwin1111) [reply src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:396]: check qbo-app

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 34]
```diff
@@ -0,0 +1,459 @@
...
+      valueField.value1 = formatDate(
+        valueField.value1,
+        'yyyy-MM-dd hh:mm:ss+hh',
+        'en'
+      );
+    }
+
+    if (typeof valueField.value1 === 'string') {
+      valueField.value1 = [valueField.value1];
+    }
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:131]: make a common function out of this, send everything needed as args

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 35]
```diff
@@ -0,0 +1,459 @@
...
+    const payload1 = {
+      condition: valueField.condition1.field_name,
+      operator: valueField.operator1,
+      values: valueField.value1,
+      rank: 1,
+      join_by: valueField.join_by ? valueField.join_by : null,
+      is_custom: valueField.condition1.is_custom,
+      custom_field_type: valueField.condition1.is_custom
+        ? valueField.condition1.type
+        : null,
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:142]: do it in a single line

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 36]
```diff
@@ -0,0 +1,459 @@
...
+        rank: 2,
+        join_by: null,
+        is_custom: valueField.condition2.is_custom,
+        custom_field_type: valueField.condition2.is_custom
+          ? valueField.condition2.type
+          : null,
+      };
+      this.settingsService
+        .postSkipExport(that.workspaceId, payload2)
+        .subscribe((skipExport: SkipExport) => {});
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:179]: remove this from here

File: src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts [hunk 37]
```diff
@@ -0,0 +1,459 @@
...
+    }
+
+    this.settingsService
+    .postSkipExport(that.workspaceId, payload1)
+    .subscribe((skipExport: SkipExport) => {
+      this.snackBar.open('Skip Export fields saved successfully');
```
Comment (ashwin1111) [inline src/app/netsuite/settings/netsuite-configurations/skip-export/skip-export.component.ts:185]: check if 2nd row exist and send another API here

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**